### PR TITLE
Update SWIG versions and add Fortran fork

### PR DESCRIPTION
**Pull request by sethrj**
- SWIG versions use commit hashes from tags at https://github.com/swig/swig/releases
- Since git repo doesn't include configure.sh, call autoreconf script
- For information on the Fortran fork: see https://arxiv.org/abs/1904.02546

### CODE REVIEW

**Comment by sethrj** [comment]
@eschnett Could you review this PR? Thanks!

File: var/spack/repos/builtin/packages/swig/package.py [hunk 1]
```diff
@@ -20,22 +20,31 @@ class Swig(AutotoolsPackage):
...
-    version('3.0.12', '82133dfa7bba75ff9ad98a7046be687c')
-    version('3.0.11', '13732eb0f1ab2123d180db8425c1edea')
-    version('3.0.10', 'bb4ab8047159469add7d00910e203124')
-    version('3.0.8',  'c96a1d5ecb13d38604d7e92148c73c97')
-    version('3.0.2',  '62f9b0d010cef36a13a010dc530d0d41')
-    version('2.0.12', 'c3fb0b2d710cc82ed0154b91e43085a4')
-    version('2.0.2',  'eaf619a4169886923e5f828349504a29')
-    version('1.3.40', '2df766c9e03e02811b1ab4bba1c7b9cc')
+    version('4.0.0', commit='9c3faeb0ee6c588bc4772bd186554796cc1e446e',
+            preferred=True)
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/swig/package.py]: `preferred=True` is not needed here, Spack already defaults to the latest numbered release.

File: var/spack/repos/builtin/packages/swig/package.py [hunk 2]
```diff
@@ -20,22 +20,31 @@ class Swig(AutotoolsPackage):
...
+    version('2.0.12', commit='18c9eca1827e009c66e5651145aff17745b76140')
+    version('1.3.40', commit='fc52ca3b80434eee001745e93c07cbf0fb7095cf')
 
     depends_on('pcre')
 
+    # Repository does *not* include configure, so it calls autoreconf
+    depends_on('m4')
+    depends_on('autoconf')
+    depends_on('automake')
+    depends_on('libtool')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/swig/package.py]: Why change to the git repo? Did the release tarballs not build? You can have a `fortran` version that downloads from git and other versions that download via URL.

**Comment by sethrj** [comment]
@adamjstewart Since someone else already updated the SWIG 4.0.0 version, I've simply *added* the git versions as options. As with other packages, it'll only `autoreconf` for the git versions, which don't have `configure` committed to them.

File: var/spack/repos/builtin/packages/swig/package.py [hunk 3]
```diff
@@ -31,12 +31,29 @@ class Swig(AutotoolsPackage):
...
     def create_symlink(self):
+        # CMake compatibility: see https://github.com/spack/spack/pull/6240
         with working_dir(self.prefix.bin):
-            os.symlink('swig', 'swig%i.0' % self.spec.version[0])
+            os.symlink('swig', 'swig{0}'.format(self.spec.version.up_to(2)))
+
+    for _version in ['@fortran', '@master']:
+        @when(_version)
+        def autoreconf(self, spec, prefix):
+            which('sh')('./autogen.sh')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/swig/package.py]: Does the default autoreconf method not work?

File: var/spack/repos/builtin/packages/swig/package.py [hunk 4]
```diff
@@ -31,12 +31,29 @@ class Swig(AutotoolsPackage):
     version('2.0.12', 'c3fb0b2d710cc82ed0154b91e43085a4')
     version('2.0.2',  'eaf619a4169886923e5f828349504a29')
     version('1.3.40', '2df766c9e03e02811b1ab4bba1c7b9cc')
+    version('master',
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/swig/package.py]: Can you move `master` above all of the stable releases? We like to sort from newest to oldest.

File: var/spack/repos/builtin/packages/swig/package.py (hunk 3 above)
Comment (sethrj) [reply var/spack/repos/builtin/packages/swig/package.py]: No, unfortunately it doesn't.

File: var/spack/repos/builtin/packages/swig/package.py (hunk 4 above)
Comment (sethrj) [reply var/spack/repos/builtin/packages/swig/package.py]: Sure thing, but I thought that would make it the 'preferred' version, whereas I would think the latest officially released version should be preferred.

Should the fortran fork also be above the 'released' versions?

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/swig/package.py]: > but I thought that would make it the 'preferred' version

The preferred version is always the newest numerical release, unless `preferred=True` is added. The order of versions in the package does not affect concretization.

> Should the fortran fork also be above the 'released' versions?

I would probably leave that at the bottom. Spack sorts all non-develop/master versions as older than every other version.

Comment (sethrj) [reply var/spack/repos/builtin/packages/swig/package.py]: Got it. Thanks for the clarification!

**Pull request merged**